File: frontend/serve.js
import http from 'node:http';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PORT = 5173;

const ROOT = path.join(__dirname, '..');
const MAIN_DIST = path.join(__dirname, 'dist');
const AI_TEST_DIST = path.join(ROOT, 'ai-test-frontend', 'dist');

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'application/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.ico': 'image/x-icon',
  '.woff': 'font/woff',
  '.woff2': 'font/woff2',
  '.ttf': 'font/ttf',
  '.eot': 'application/vnd.ms-fontobject',
};

// [path_prefix, target_url]
const API_PROXY_RULES = [
  ['/api/ai/speaking/scoring', 'http://localhost:5181'],
  ['/api/ai/speaking', 'http://localhost:5181'],
  ['/api/admin/speaking', 'http://localhost:5181'],
  ['/api/ai/writing', 'http://localhost:5182'],
  ['/api/ai/evaluation', 'http://localhost:5182'],
  ['/api/admin/ai', 'http://localhost:5182'],
  ['/api/ai/import', 'http://localhost:5186'],
  ['/api/agent', 'http://localhost:5187'],
];

const BACKEND = 'http://localhost:8080';

const proxyAgent = new http.Agent({ keepAlive: true, maxSockets: 100, maxFreeSockets: 10 });

function proxyTo(url, target, req, res) {
  const targetUrl = new URL(target);
  const options = {
    method: req.method,
    hostname: targetUrl.hostname,
    port: targetUrl.port,
    path: url.pathname + url.search,
    headers: { ...req.headers, host: targetUrl.host },
    agent: proxyAgent,
  };

  const proxyReq = http.request(options, (proxyRes) => {
    res.writeHead(proxyRes.statusCode, proxyRes.headers);
    proxyRes.pipe(res);
  });

  proxyReq.on('error', () => {
    if (!res.destroyed && !res.headersSent) {
      res.writeHead(504, { 'Content-Type': 'text/plain; charset=utf-8' });
      res.end('504 Gateway Timeout - Backend unavailable');
    }
  });

  res.on('close', () => {
    if (!res.writableFinished) proxyReq.destroy();
  });

  req.pipe(proxyReq);
}

function serveStatic(res, filePath, fallbackHtml) {
  const ext = path.extname(filePath);
  const contentType = MIME[ext] || 'application/octet-stream';

  fs.readFile(filePath, (err, data) => {
    if (err) {
      if (err.code === 'ENOENT' && fallbackHtml) {
        return fs.readFile(fallbackHtml, (err2, data2) => {
          if (err2) {
            res.writeHead(404);
            return res.end('Not Found');
          }
          res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
          res.end(data2);
        });
      }
      res.writeHead(500);
      return res.end('Internal Server Error');
    }

    const headers = { 'Content-Type': contentType };
    if (ext === '.js' || ext === '.css') {
      headers['Cache-Control'] = 'public, max-age=31536000, immutable';
    }
    res.writeHead(200, headers);
    res.end(data);
  });
}

const server = http.createServer((req, res) => {
  const url = new URL(req.url, `http://${req.headers.host || 'localhost'}`);

  // AI API proxy
  for (const [prefix, target] of API_PROXY_RULES) {
    if (url.pathname.startsWith(prefix)) {
      return proxyTo(url, target, req, res);
    }
  }

  // Backend API proxy
  if (url.pathname.startsWith('/api/')) {
    return proxyTo(url, BACKEND, req, res);
  }

  // AI Test UI at /ai-test/
  if (url.pathname.startsWith('/ai-test')) {
    const basePath = '/ai-test';
    const relativePath = url.pathname === basePath || url.pathname === basePath + '/'
      ? '/index.html'
      : url.pathname.slice(basePath.length);
    const filePath = path.join(AI_TEST_DIST, relativePath);
    const fallback = path.join(AI_TEST_DIST, 'index.html');
    return serveStatic(res, filePath, fallback);
  }

  // Main frontend
  const filePath = path.join(MAIN_DIST, url.pathname === '/' ? 'index.html' : url.pathname);
  const fallback = path.join(MAIN_DIST, 'index.html');
  serveStatic(res, filePath, fallback);
});

server.keepAliveTimeout = 5000;
server.timeout = 0;

server.listen(PORT, '127.0.0.1', () => {
  console.log(`[serve] Production server running on http://127.0.0.1:${PORT}`);
  console.log(`[serve] Main FE: ${MAIN_DIST}`);
  console.log(`[serve] AI Test UI: ${AI_TEST_DIST} -> /ai-test/`);
  console.log(`[serve] Proxying AI APIs -> Python services (5182, 5181)`);
  console.log(`[serve] Proxying /api/* -> ${BACKEND}`);
});

File: frontend/serve.py
import asyncio
import mimetypes
import os
import re
import subprocess
from pathlib import Path

import httpx

DIST_DIR = Path(__file__).parent / "dist"
API_TARGET = "http://localhost:8080"
AGENT_TARGET = "http://localhost:5187"
PORT = 5173


async def handle_client(reader, writer):
    request_line = await reader.readline()
    if not request_line:
        writer.close()
        return

    method, path, _ = request_line.decode().strip().split(" ", 2)
    headers = {}
    while True:
        line = await reader.readline()
        if line == b"\r\n":
            break
        key, _, value = line.decode().strip().partition(": ")
        headers[key.lower()] = value

    content_length = int(headers.get("content-length", 0))
    body = await reader.readexactly(content_length) if content_length > 0 else b""

    if re.match(r"^/api/agent/sessions/\d+/stream$", path):
        await proxy_sse_direct(method, path, headers, writer)
    elif path.startswith("/api/"):
        await proxy_api(method, path, headers, body, writer)
    elif path.startswith("/uploads/"):
        await proxy_uploads(method, path, headers, writer)
    else:
        await serve_static(path, writer)


_FILE_EXT = re.compile(r"\.[a-zA-Z0-9]{2,5}(?:\?.*)?$")


async def serve_static(path, writer):
    clean_path = path.split("?")[0].split("#")[0]

    if clean_path == "/":
        clean_path = "/index.html"

    file_path = DIST_DIR / clean_path.lstrip("/")
    if file_path.exists() and file_path.is_file():
        content_type, _ = mimetypes.guess_type(str(file_path))
        if content_type is None:
            content_type = "application/octet-stream"

        content = file_path.read_bytes()
        resp = (
            f"HTTP/1.1 200 OK\r\n"
            f"Content-Type: {content_type}\r\n"
            f"Content-Length: {len(content)}\r\n"
            f"Cache-Control: public, max-age=3600\r\n"
            f"Access-Control-Allow-Origin: *\r\n"
            f"\r\n"
        ).encode() + content
        writer.write(resp)
    elif _FILE_EXT.search(clean_path):
        writer.write(b"HTTP/1.1 404 Not Found\r\nContent-Length: 0\r\n\r\n")
    else:
        # SPA fallback: serve index.html
        index_path = DIST_DIR / "index.html"
        if index_path.exists():
            content = index_path.read_bytes()
            resp = (
                f"HTTP/1.1 200 OK\r\n"
                f"Content-Type: text/html; charset=utf-8\r\n"
                f"Content-Length: {len(content)}\r\n"
                f"\r\n"
            ).encode() + content
            writer.write(resp)
        else:
            writer.write(b"HTTP/1.1 404 Not Found\r\nContent-Length: 0\r\n\r\n")

    await writer.drain()
    writer.close()


async def proxy_uploads(method, path, headers, writer):
    """Proxy /uploads/ to Python agent (image files)."""
    try:
        async with httpx.AsyncClient(timeout=30) as client:
            resp = await client.get(f"{AGENT_TARGET}{path}")
            content_type = resp.headers.get("content-type", "image/jpeg")
            writer.write(
                f"HTTP/1.1 {resp.status_code} OK\r\n"
                f"Content-Type: {content_type}\r\n"
                f"Content-Length: {len(resp.content)}\r\n"
                f"Cache-Control: public, max-age=86400\r\n"
                f"Access-Control-Allow-Origin: *\r\n"
                f"\r\n".encode() + resp.content
            )
    except Exception as e:
        writer.write(f"HTTP/1.1 404 Not Found\r\nContent-Length: 0\r\n\r\n".encode())
    await writer.drain()
    writer.close()


async def proxy_sse_direct(method, path, headers, writer):
    """Proxy SSE stream directly to Python agent, bypass Java."""
    try:
        async with httpx.AsyncClient(timeout=None) as client:
            req_headers = {"Accept": "text/event-stream"}
            if "authorization" in headers:
                req_headers["Authorization"] = headers["authorization"]

            async with client.stream(method, f"{AGENT_TARGET}{path}", headers=req_headers) as resp:
                writer.write(
                    f"HTTP/1.1 {resp.status_code} OK\r\n"
                    f"Content-Type: text/event-stream\r\n"
                    f"Cache-Control: no-cache\r\n"
                    f"Connection: keep-alive\r\n"
                    f"Access-Control-Allow-Origin: *\r\n"
                    f"\r\n".encode()
                )
                async for chunk in resp.aiter_bytes():
                    try:
                        writer.write(chunk)
                        await writer.drain()
                    except Exception:
                        break
    except Exception as e:
        try:
            error = f'data: {{"type":"error","error":"{str(e)}"}}\n\n'
            writer.write(f"HTTP/1.1 502 Bad Gateway\r\nContent-Type: text/event-stream\r\n\r\n{error}".encode())
            await writer.drain()
        except Exception:
            pass
    finally:
        try:
            writer.close()
        except Exception:
            pass


async def proxy_api(method, path, headers, body, writer):
    try:
        async with httpx.AsyncClient(timeout=120) as client:
            target_headers = {
                "Content-Type": headers.get("content-type", "application/json"),
                "Accept": "text/event-stream,application/json",
            }
            if "authorization" in headers:
                target_headers["Authorization"] = headers["authorization"]

            resp = await client.request(
                method=method,
                url=f"{API_TARGET}{path}",
                headers=target_headers,
                content=body,
            )

            if "text/event-stream" in resp.headers.get("content-type", ""):
                writer.write(
                    f"HTTP/1.1 {resp.status_code} OK\r\n"
                    f"Content-Type: text/event-stream\r\n"
                    f"Cache-Control: no-cache\r\n"
                    f"Connection: keep-alive\r\n"
                    f"Access-Control-Allow-Origin: *\r\n"
                    f"\r\n".encode()
                )
                async for chunk in resp.aiter_bytes():
                    writer.write(chunk)
                    await writer.drain()
            else:
                response_headers = (
                    f"HTTP/1.1 {resp.status_code} {'OK' if resp.status_code < 400 else 'Error'}\r\n"
                    f"Content-Type: {resp.headers.get('content-type', 'application/json')}\r\n"
                    f"Content-Length: {len(resp.content)}\r\n"
                    f"Access-Control-Allow-Origin: *\r\n"
                    f"\r\n"
                )
                writer.write(response_headers.encode() + resp.content)

    except Exception as e:
        error = f'{{"error": "Proxy error: {str(e)}"}}'.encode()
        writer.write(
            f"HTTP/1.1 502 Bad Gateway\r\n"
            f"Content-Type: application/json\r\n"
            f"Content-Length: {len(error)}\r\n"
            f"\r\n".encode() + error
        )

    await writer.drain()
    writer.close()


async def main():
    server = await asyncio.start_server(handle_client, "0.0.0.0", PORT)
    addr = server.sockets[0].getsockname()
    print(f"Serving frontend from {DIST_DIR} on http://{addr[0]}:{addr[1]}")
    print(f"API proxy to {API_TARGET}")

    async with server:
        await server.serve_forever()


if __name__ == "__main__":
    asyncio.run(main())
